src/Http/Request.php

```diff
@@ -0,0 +1,332 @@
+
+  /**
+   * The input HTTP headers.
+   *
+   * @var HttpHeaderBag
+   */
+  protected $headers;
+
+  /**
+   * The unprocessed body of the request.
+   *
+   * This should be a PHP stream, but let's keep it simple.
+   *
+   * @var string
+   */
+  protected $body;
+
+  /**
+   * Indicates if the request was routed by the menu system.
+   *
+   * @var bool
+   */
+  protected $viaRouter = FALSE;
+
+  /**
+   * The passed in CSRF token in the corresponding header.
+   *
+   * @var string
+   */
+  protected $csrfToken = NULL;
+
+  /**
+   * Cookies for the request.
+   *
+   * @var array
+   */
+  protected $cookies = array();
+
+  /**
+   * Files attached to the request.
+   *
+   * @var array
+   */
+  protected $files = array();
+
+  /**
+   * Server information.
+   *
+   * @var array
+   */
+  protected $server = array();
+
+  /**
+   * Holds the parsed body.
+   *
+   * @internal
+   * @var \ArrayObject
+   */
+  private $parsedBody;
+
+  /**
+   * Constructor.
+   *
+   * Parses the URL and the query params. It also uses input:// to get the body.
+   */
+  public function __construct($path, $query, $method = 'GET', HttpHeaderBag $headers, $via_router = FALSE, $csrf_token = NULL, $cookies = array(), $files = array(), $server = array()) {
+    $this->path = $path;
+    $this->query = $query;
+    $this->method = $method;
+    $this->headers = $headers;
+    $this->viaRouter = $via_router;
+    $this->csrfToken = $csrf_token;
+    $this->cookies = $cookies;
+    $this->files = $files;
+    $this->server = $server;
+
+    // Allow implementing modules to alter the request.
+    drupal_alter('restful_parse_request', $this);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create($path, $query, $method = 'GET', HttpHeaderBag $headers, $via_router = FALSE, $csrf_token = NULL, $cookies = array(), $files = array(), $server = array()) {
+    if ($method == static::METHOD_POST && $headers->get('x-http-method-override')) {
+      $method = $headers->get('x-http-method-override')->contents();
+    }
+    if (!static::isValidMethod($method, FALSE)) {
+      throw new \RestfulBadRequestException('Unrecognized HTTP method.');
+    }
+    return new static($path, $query, $method, $headers, $via_router, $csrf_token, $cookies, $files, $server);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createFromGlobals() {
+    $path = implode('/', arg());
+    $query =  drupal_get_query_parameters();
+    $method = strtoupper($_SERVER['REQUEST_METHOD']);
+
+    // This flag is used to identify if the request is done "via Drupal" or "via
+    // CURL";
+    $via_router = TRUE;
+    $headers = static::parseHeadersFromGlobals();
+    $csrf_token = $headers->get('x-csrf-Token')->contents();
+
+    return static::create($path, $query, $method, $headers, $via_router, $csrf_token, $_COOKIE, $_FILES, $_SERVER);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function isWriteMethod($method, $strict = TRUE) {
```

Comment:
I don't remember why we have the $strict, I think it had to do something with testing -- can we remove it?